[Workplace Search] Add Account Settings page imported from Security plugin

yakhinvadim opened the pull request:
This description is outdated, skip to https://github.com/elastic/kibana/pull/99791#issuecomment-853090859 for latest update.

## Summary

This PR adds the Account Settings page to Workplace Search personal dashboard. The components for this page were imported from the Security plugin.

@azasypkin 
I'm looking for feedback on the general approach. Please let me know what you think and if you see any areas for improvement. Commits https://github.com/elastic/kibana/pull/99791/commits/632589f9a2f456e15754aa1d1e443bd74dbbbef2 and https://github.com/elastic/kibana/pull/99791/commits/3a1768172f9dd4f990a9c3f362eb6e72b7e189ad should be most relevant to you.

 I will request a review from my team if the general approach looks good to you. 

---

### Screenshots

Workplace Search Account Settings page:

https://user-images.githubusercontent.com/11838280/117842838-b0509480-b254-11eb-94dd-c6ffb2e7de2c.mp4

Current Account page in Security plugin for the reference:

![image](https://user-images.githubusercontent.com/11838280/117843584-51d7e600-b255-11eb-98cd-20d84d83ba2d.png)

---

### Checklist

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125561/)
* Commit: 2c350dc99b4f5d8f01718141cd882d57fa1ee96d
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-99791/2c350dc99b4f5d8f01718141cd882d57fa1ee96d)
* This comment will update when the build is complete

### Failed CI Steps
* [Execute xpack-kibana-ciGroup9](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125561/execution/node/857/log)
* [Execute xpack-kibana-ciGroup9](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125561/execution/node/1087/log)

### History
* :broken_heart: [Build #125333](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125333/) failed 48d7bfa568a395f9d405d347d40c1fffdeaf623d
* :broken_heart: [Build #125314](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125314/) failed 3c87b4b4a44a14c2478323c4e52aa00776e46b26

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125314/","number":"125314","commit":"3c87b4b4a44a14c2478323c4e52aa00776e46b26"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125333/","number":"125333","commit":"48d7bfa568a395f9d405d347d40c1fffdeaf623d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125561/","number":"125561","commit":"2c350dc99b4f5d8f01718141cd882d57fa1ee96d"}]}
PIPELINE-->

yakhinvadim commented:
Thanks again for your help with this, Oleg! 

I have several questions:

1. Page load bundle increase for `security` plugin worries me a bit. Is this something we can keep at the previous level or is the increase unavoidable?
1. The test errors are coming from `ui_capabilities/spaces_only` test suite, but I didn't touch it. Do you have any insights into why the tests are failing?
1. This is the first time I see `Public APIs missing comments` and `Public APIs missing exports` warnings. Should I run the recommended commands and commit the diff in this PR?

Questions 2 and 3 are not directly related to security plugin, so please let me know if I should seek help elsewhere!

azasypkin reviewed (commented):
The approach looks good to me. We all will probably need to sync again in the future when we decide to update how personal info screen looks like (once we have user personal settings, notifications and whatnot), but that sounds like a reasonable price for reusable components. 

@legrego do see any problems with sharing `ChangePassword` and `PersonalInfo` components with other plugins that I'm missing? Here is some context: the personal dashboards for Workplace Search will be part of the Workplace Search Kibana plugin, or more precisely part of the a *chromeless* app in the Kibana plugin. Currently such dashboards contain a form to change the user password that Workplace Search users are already familiar with. Ideally we'd integrate Kibana own account management view into this flow somehow, but it doesn't seem like we have a good (architecturally and UX-wise) way of doing this now. That's how the idea to reuse only certain security UI components came up. 

> 1. Page load bundle increase for security plugin worries me a bit. Is this something we can keep at the previous level, or is the increase unavoidable?

Yeah, that worries me a bit too, but I'm not sure if we can do anything about it unless we do something like this in Security plugin:
```ts

import type { ChangePassword } from './account_management/change_password';
import type { PersonalInfo } from './account_management/personal_info';
export type { ChangePassword, PersonalInfo };

export function loadUserComponents(): Promise<{
  ChangePassword: typeof ChangePassword;
  PersonalInfo: typeof PersonalInfo;
}> {
  return Promise.all([
    import('./account_management/change_password'),
    import('./account_management/personal_info'),
  ]).then(([comp1, comp2]) => ({
    ChangePassword: comp1.ChangePassword,
    PersonalInfo: comp2.PersonalInfo,
  }));
}
```

@elastic/kibana-operations is there any recommended way to export `public` components from plugin A, so that plugin B can import them, but these components aren't included into main bundle of the plugin A?

> 2. The test errors are coming from the ui_capabilities/spaces_only test suite, but I didn't touch it. Do you have any insights into why the tests are failing?

Yeah, as you see only `spaces_only` tests are failing. These tests are run with disabled security, and since you made security required dependency for EntSearch, your plugins are disabled in these tests and don't register their specific UI capabilities anymore. As I asked you in another comment, I'm not sure if it's what you really want. If it's intentional, then just remove capabilities your app registers from the test expected results.

> 3. This is the first time I see Public APIs missing comments and Public APIs missing exports warnings. Should I run the recommended commands and commit the diff in this PR?

It's the first time I see this as well, but if running these commands produces changes that make CI happy, please do that. If not, let me know and I'll figure this out.

legrego commented:
@azasypkin thanks for the ping and the context.

> Page load bundle increase for security plugin worries me a bit. Is this something we can keep at the previous level, or is the increase unavoidable?

I'm not a fan of the increased bundle size either -- while we are technically under the `100kb` recommended limit, we spent a lot of time optimizing this bundle recently, so I'd like to keep this as small as possible.

We took a different approach with the Spaces plugin recently (https://github.com/elastic/kibana/pull/91976), in order to share reusable components between plugins. Essentially the components are available through an async function on the plugin's lifecycle methods, which allows us to lazy-load them, which in turn keeps them out of the page load bundle. It's a bit more work, and less elegant...but it's the best we've come up with so far. I'd suggest taking a similar approach here if we really need these components to be sharable.


> do see any problems with sharing ChangePassword and PersonalInfo components with other plugins

I have some questions first -- they are probably silly questions, but I want to make sure I understand the problem we're trying to solve here.

1) Perhaps this is better suited for the product team, but why does this chromeless application lack the global Elastic header, which already includes the account nav? I was under the impression that this was more or less a permanent fixture, allowing folks to navigate throughout Elastic (Cloud Console, support portal, etc). Is the goal of this chromeless app to hide the rest of Kibana, or to hide the rest of Elastic?
2) What is the expected experience for our Cloud users? Our cloud users will be using SSO more and more in the future, and these components aren't all that useful unless you happen to authenticate via the `native` realm. Cloud effectively replaces our default "Profile" link to send users to their Cloud profile instead, since that is arguably more useful to them. Our components are still available to Cloud users, just under a different link. See https://github.com/elastic/kibana/pull/97870 for sample screenshots.

yakhinvadim commented:
@legrego These are great questions, thanks for asking them!

## Question 1

> Perhaps this is better suited for the product team, but why does this chromeless application lack the global Elastic header, which already includes the account nav? I was under the impression that this was more or less a permanent fixture, allowing folks to navigate throughout Elastic (Cloud Console, support portal, etc). Is the goal of this chromeless app to hide the rest of Kibana, or to hide the rest of Elastic?

I'll copy the content from a product document that has a good explanation:

---

> ### Summary
> 
> - Enterprise Search is dropping support for its standalone UIs with 8.0, and will rely exclusively on Kibana as its user interface for experience management.
> - Workplace Search’s _search user_ persona represents a **new type of user for the Kibana ecosystem**, and certain needs specific to this new user need to be addressed.
> - A pared down version of Kibana is most appropriate for search users.
> 
> ### Defining the Search User persona
> 
> Workplace Search brings a unique persona to the Elastic ecosystem: the search (end-)user persona. Search users are non-operators, and their experience relies on the Stack and its operational functionality being entirely obfuscated and abstracted. Here are a few key concepts to understand:
> 
> - Search users are first-class users in the ecosystem, represented as Elasticsearch users, with specific roles (or role mappings) associated.
> - Search users are not operators or administrators.
> - Search users are assumed to be non-technical users, with no explicit knowledge of a work tool’s underlying technology layers.
> - Search users are cognitively trained to use a particular product (in this case Workplace Search) and not a platform or technology (i.e. Kibana, Elastic or Elasticsearch).
> - Search users require some level of personal account management from a distinct management area that is non-technical in nature. This includes consuming personal analytics and connecting private data sources.
> 
> 
> EXAMPLE
> **Amanda**, who contributes to the Customer Success team at Acme Co., has access to the Workplace Search experience where she can easily query the data she has access to using a standalone search interface. She uses the Workplace Search credentials that were provided to her by her IT department. 
> 
> Amanda may also want to add Slack as a Content Source for her search experience, for which she requires a simple management experience. This management experience is _powered_ by Kibana, but **does not present** the traditional operator-specific characteristics usually represented in Kibana.
> 
> Visually and aspirationally speaking, the experience for Amanda should look as follows at a high-level:
> ![unnamed (11)](https://user-images.githubusercontent.com/11838280/118690008-01750100-b7de-11eb-8ee8-0dc108f0007a.png)
> 
> … whereas the experience for **Catherine (who’s an admin)** is aligned with the traditional Kibana experience: 
>  ![unnamed (3)](https://user-images.githubusercontent.com/11838280/118690158-25d0dd80-b7de-11eb-92a7-3653a30d8a31.png)

---

To summarize: the Workplace Search personal dashboard application hides the global Elastic header because the search user most likely doesn't know about Elastic or Kibana. The goal of this chromeless app is to hide the rest of Elastic.

Hope this clarifies things a bit. Let me know if I could provide more detail.

Note: We might bring the Elastic header back somewhere in 8.x+, but for 7.x the decision was to hide it completely.

## Question 2

> What is the expected experience for our Cloud users? Our cloud users will be using SSO more and more in the future, and these components aren't all that useful unless you happen to authenticate via the native realm. Cloud effectively replaces our default "Profile" link to send users to their Cloud profile instead, since that is arguably more useful to them. Our components are still available to Cloud users, just under a different link. See #97870 for sample screenshots.

Good question. I don't have a definitive answer right now. I expect that we will either:

1) hide Workplace Search Account Settings and link to it completely for such users, or
2) show them a not that useful version of this page.

This will require a discussion with the product team.

For now not having a polished experience is expected and ok, because the Workplace Search Kibana plugin will be in beta until 8.0.

## Bundle size increase

Thank you for the link! I'll look into loading components through an async function.

yakhinvadim commented on x-pack/plugins/enterprise_search/kibana.json:
```diff
@@ -2,9 +2,9 @@
   "version": "kibana",
   "kibanaVersion": "kibana",
-  "requiredPlugins": ["features", "licensing", "charts"],
+  "requiredPlugins": ["features", "licensing", "charts", "security"],
```
Great catch, I haven't considered this case. I'll look into it. 
(Currently, I think I'll make it an optional dependency again).

legrego commented:
@yakhinvadim thanks so much for the context, that's very helpful. Overall I don't have any objections to the approach you've outlined above for reusing our components, so long as we can address the async loading to keep the bundle sizes small.

I'm satisfied with your answers to my first question, and I agree that we need a discussion with the product team about the second question.

Ping @alexfrancoeur since we worked together recently on the profile links for cloud users 🙂. Workplace Search is adding an inlined version of Kibana's account settings page within their personal dashboard view. This will work well for users in the `native` realm, but we don't know what the experience should look like for cloud users who come in via SSO. We can show these same controls for now, but they won't be all that useful. Also worth noting that this interface is beta for the remainder of `7.x`, so we have some time to decide if we wish to leave it as-is and defer the decision.

spalger commented:
> @elastic/kibana-operations is there any recommended way to export `public` components from plugin A, so that plugin B can import them, but these components aren't included into main bundle of the plugin A?

The only real option is to export an async function from the root of plugin A and then use a `return async import()` in that async function, which you will need to call from B.

We could probably add the ability to specify an `extraAsyncPluginDirs` or something in `kibana.json` which would only be accessible with `async import()` statements, but it will still need to run the import in an async context somewhere so a `loadDependencyFromPluginA()` function works just as well in my eyes.

The holy grail would be to allow standard `import` statements in an async chunk of plugin B to load an async chunk from plugin A, but implementing that would take a good deal of time and doesn't seem to have enough demand to justify it right now.

alexfrancoeur commented:
> Ping @alexfrancoeur since we worked together recently on the profile links for cloud users 🙂. Workplace Search is adding an inlined version of Kibana's account settings page within their personal dashboard view. This will work well for users in the native realm, but we don't know what the experience should look like for cloud users who come in via SSO. We can show these same controls for now, but they won't be all that useful. Also worth noting that this interface is beta for the remainder of 7.x, so we have some time to decide if we wish to leave it as-is and defer the decision.

Thanks for the ping @legrego! For the time being, until we have a better user model for the Cloud / Stack, I'm wondering if we should be providing a similar experience that we do for Kibana. Do we have a sense of how many users will be coming from Cloud SSO? @jonasll @yakhinvadim Let me know if it'd be helpful to connect for 15 minutes and talk through what that experience is today.

jonasll commented:
@alexfrancoeur It would be helpful.

To answer the question: I expect few users (not admins) to come from Cloud SSO or native. We expect most to come from organizational SSO experiences powered by AD, Okta and the likes.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128501/)
* Commit: efa7d683110ce4bd9b6533e1eddf7bf17f9a59ed
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-99791/efa7d683110ce4bd9b6533e1eddf7bf17f9a59ed)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/4aa3920746645ac65e25e732bfaf2645c59b211e) | [after](https://github.com/elastic/kibana/commit/efa7d683110ce4bd9b6533e1eddf7bf17f9a59ed) | diff |
| --- | --- | --- | --- |
| `enterpriseSearch` | 1398 | 1400 | +2 |
| `security` | 475 | 482 | +7 |
| total |  |  | +9 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/4aa3920746645ac65e25e732bfaf2645c59b211e) | [after](https://github.com/elastic/kibana/commit/efa7d683110ce4bd9b6533e1eddf7bf17f9a59ed) | diff |
| --- | --- | --- | --- |
| `enterpriseSearch` | 2.1MB | 2.1MB | +2.5KB |
| `security` | 765.0KB | 780.5KB | +15.6KB |
| total |  |  | +18.1KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/4aa3920746645ac65e25e732bfaf2645c59b211e) | [after](https://github.com/elastic/kibana/commit/efa7d683110ce4bd9b6533e1eddf7bf17f9a59ed) | diff |
| --- | --- | --- | --- |
| `security` | 15 | 16 | +1 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/4aa3920746645ac65e25e732bfaf2645c59b211e) | [after](https://github.com/elastic/kibana/commit/efa7d683110ce4bd9b6533e1eddf7bf17f9a59ed) | diff |
| --- | --- | --- | --- |
| `security` | 82.3KB | 85.2KB | +2.8KB |

<details><summary>Unknown metric groups</summary>

#### async chunk count
| id | [before](https://github.com/elastic/kibana/commit/4aa3920746645ac65e25e732bfaf2645c59b211e) | [after](https://github.com/elastic/kibana/commit/efa7d683110ce4bd9b6533e1eddf7bf17f9a59ed) | diff |
| --- | --- | --- | --- |
| `security` | 21 | 23 | +2 |

#### References to deprecated APIs
| id | [before](https://github.com/elastic/kibana/commit/4aa3920746645ac65e25e732bfaf2645c59b211e) | [after](https://github.com/elastic/kibana/commit/efa7d683110ce4bd9b6533e1eddf7bf17f9a59ed) | diff |
| --- | --- | --- | --- |
| `canvas` | 29 | 25 | -4 |
| `crossClusterReplication` | 8 | 6 | -2 |
| `fleet` | 22 | 20 | -2 |
| `globalSearch` | 4 | 2 | -2 |
| `indexManagement` | 12 | 7 | -5 |
| `infra` | 261 | 149 | -112 |
| `lens` | 67 | 45 | -22 |
| `licensing` | 18 | 15 | -3 |
| `lists` | 239 | 236 | -3 |
| `maps` | 286 | 208 | -78 |
| `ml` | 121 | 115 | -6 |
| `monitoring` | 109 | 56 | -53 |
| `securitySolution` | 390 | 346 | -44 |
| `stackAlerts` | 101 | 95 | -6 |
| total |  |  | -342 |


</details>

### History
* :broken_heart: [Build #125561](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125561/) failed 2c350dc99b4f5d8f01718141cd882d57fa1ee96d
* :broken_heart: [Build #125333](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125333/) failed 48d7bfa568a395f9d405d347d40c1fffdeaf623d
* :broken_heart: [Build #125314](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125314/) failed 3c87b4b4a44a14c2478323c4e52aa00776e46b26

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125314/","number":"125314","commit":"3c87b4b4a44a14c2478323c4e52aa00776e46b26"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125333/","number":"125333","commit":"48d7bfa568a395f9d405d347d40c1fffdeaf623d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125561/","number":"125561","commit":"2c350dc99b4f5d8f01718141cd882d57fa1ee96d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128501/","number":"128501","commit":"efa7d683110ce4bd9b6533e1eddf7bf17f9a59ed"}]}
PIPELINE-->

yakhinvadim commented:
@azasypkin This should be ready for re-review. 

The first 4 commits in this PR only touch Security plugin files. Last 4 - only Enterprise Search plugin files. I will request a review from my team for Enterprise Search changes after you approve the changes in the Security plugin.

### Major changes since last time

- Components `PersonalInfo` and `ChangePassword` are now available via the `uiApi` property in Security plugin lifecycle methods. The effect on the Security plugin's page load bundle size is minor: +2.8KB.
- The Security plugin is now optional in the Enterprise Search solution (consists of 2 products: App Search and Workplace Search):
  - If the Security plugin is **enabled**, the Workplace Search's Account Settings page asynchronously loads and shows the` PersonalInfo` and `ChangePassword` components.
  - If the Security plugin is **disabled**, only the App Search product is available. Workplace Search and the Account Settings page become unavailable. `PersonalInfo` and `ChangePassword` components are not loaded.

### Implementation details

- I mostly copied the patterns from the Spaces plugin (see https://github.com/elastic/kibana/pull/99791#issuecomment-841413210) with minimal changes for two reasons: 1) this works and has been approved before and 2) it should be easier to DRY out in the future if needed.

### Screencast

Throttled to Fast 3G. Can be skipped to 0:15.

https://user-images.githubusercontent.com/11838280/120501182-cbb44880-c397-11eb-92a7-c9b8dd7992fd.mp4

azasypkin commented:
__ACK:__ will review on Monday, Jun 7th at the latest.

azasypkin reviewed (commented):
Looks good, just left a few suggestions. I know you tried to replicate the lazy-loading code from the Spaces as much as possible, but I believe there is no reason to reproduce all the complexity we needed there, but may never need here.

What do you think?

yakhinvadim commented:
Thanks for the review, @azasypkin!
I agree with all the suggestions. I'll update the PR and ping you when it's ready for re-review.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/129846/)
* Commit: 140cd91fc7ea4717be957322a215833d0f9ba368
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-99791/140cd91fc7ea4717be957322a215833d0f9ba368)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/451f64bde0a5ceb80137f0e3fd2e73d9969928fc) | [after](https://github.com/elastic/kibana/commit/140cd91fc7ea4717be957322a215833d0f9ba368) | diff |
| --- | --- | --- | --- |
| `enterpriseSearch` | 1417 | 1419 | +2 |
| `security` | 475 | 482 | +7 |
| total |  |  | +9 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/451f64bde0a5ceb80137f0e3fd2e73d9969928fc) | [after](https://github.com/elastic/kibana/commit/140cd91fc7ea4717be957322a215833d0f9ba368) | diff |
| --- | --- | --- | --- |
| `enterpriseSearch` | 2.1MB | 2.1MB | +1.7KB |
| `security` | 767.0KB | 782.6KB | +15.6KB |
| total |  |  | +17.3KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/451f64bde0a5ceb80137f0e3fd2e73d9969928fc) | [after](https://github.com/elastic/kibana/commit/140cd91fc7ea4717be957322a215833d0f9ba368) | diff |
| --- | --- | --- | --- |
| `security` | 15 | 16 | +1 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/451f64bde0a5ceb80137f0e3fd2e73d9969928fc) | [after](https://github.com/elastic/kibana/commit/140cd91fc7ea4717be957322a215833d0f9ba368) | diff |
| --- | --- | --- | --- |
| `security` | 82.3KB | 85.1KB | +2.8KB |

<details><summary>Unknown metric groups</summary>

#### async chunk count
| id | [before](https://github.com/elastic/kibana/commit/451f64bde0a5ceb80137f0e3fd2e73d9969928fc) | [after](https://github.com/elastic/kibana/commit/140cd91fc7ea4717be957322a215833d0f9ba368) | diff |
| --- | --- | --- | --- |
| `security` | 21 | 23 | +2 |


</details>

### History
* :green_heart: [Build #128501](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128501/) succeeded efa7d683110ce4bd9b6533e1eddf7bf17f9a59ed
* :broken_heart: [Build #125561](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125561/) failed 2c350dc99b4f5d8f01718141cd882d57fa1ee96d
* :broken_heart: [Build #125333](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125333/) failed 48d7bfa568a395f9d405d347d40c1fffdeaf623d
* :broken_heart: [Build #125314](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125314/) failed 3c87b4b4a44a14c2478323c4e52aa00776e46b26

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125314/","number":"125314","commit":"3c87b4b4a44a14c2478323c4e52aa00776e46b26"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125333/","number":"125333","commit":"48d7bfa568a395f9d405d347d40c1fffdeaf623d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125561/","number":"125561","commit":"2c350dc99b4f5d8f01718141cd882d57fa1ee96d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128501/","number":"128501","commit":"efa7d683110ce4bd9b6533e1eddf7bf17f9a59ed"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/129846/","number":"129846","commit":"140cd91fc7ea4717be957322a215833d0f9ba368"}]}
PIPELINE-->

yakhinvadim commented on x-pack/plugins/security/public/account_management/change_password/change_password_async.tsx:
```diff
@@ -0,0 +1,17 @@
+  return (props: ChangePasswordProps) => {
+    return <ChangePassword {...props} />;
+  };
+};
```
Was really happy it worked! See this commit: https://github.com/elastic/kibana/pull/99791/commits/9f18ea9db06e9bcbdbee51364b8987729bfe15d3

yakhinvadim commented on x-pack/plugins/security/public/ui_api/index.ts:
```diff
@@ -0,0 +1,39 @@
+    getChangePassword: LazyComponentFn<ChangePasswordProps>;
+  };
+  UserAPIClient: typeof UserAPIClient;
+}
```
Removed it here: https://github.com/elastic/kibana/pull/99791/commits/706ac2ebb455686bec1a201fd6706e66dd56f4aa

yakhinvadim commented on x-pack/plugins/security/public/ui_api/index.ts:
```diff
@@ -0,0 +1,39 @@
+
+import type { StartServicesAccessor } from 'src/core/public';
+
+import type { ChangePasswordProps } from '../account_management/change_password';
```
Update here: https://github.com/elastic/kibana/pull/99791/commits/ebc2753855b37c9a8d4fa6660d9ffcf9b6ab3f6e. Could you confirm that this is what we want? I'm still a bit confused about what should be exported from components and what from index files.

yakhinvadim commented on x-pack/plugins/security/public/plugin.tsx:
```diff
@@ -97,6 +100,10 @@ export class SecurityPlugin
       logoutUrl,
     });
 
+    this.uiApi = getUiApi({
```
Great suggestion! Updated here: https://github.com/elastic/kibana/pull/99791/commits/4f8f832880303c35490279e18c9d1cb715e46e03

yakhinvadim commented:
@azasypkin This should be ready for re-review!

azasypkin commented on x-pack/plugins/security/public/account_management/change_password/change_password_async.tsx:
```diff
@@ -0,0 +1,29 @@
+
+export const getChangePasswordComponent = async (
+  core: CoreStart
+): Promise<React.FC<Pick<ChangePasswordProps, 'user'>>> => {
```
__nit:__ I learnt yesterday that [`Pick` doesn't play well with our dev docs](https://github.com/elastic/kibana/pull/101661#discussion_r648260136). Would you mind creating two separate interfaces inside of `change_password` instead and replacing any `Pick`-based types with dedicated interface?

```ts
export interface ChangePasswordProps {
  user: AuthenticatedUser;
}

export interface ChangePasswordPropsInternal extends ChangePasswordProps {
  userAPIClient: PublicMethodsOf<UserAPIClient>;
  notifications: NotificationsSetup;
}
```

(and you also need to merge the latest master and likely to update latest version of `SecurityPluginStart`)

yakhinvadim replied:
Ah, good to know, and thanks for the link! I replaced it here: https://github.com/elastic/kibana/pull/99791/commits/24258211021c0b995ff548521a172acaac11c636

azasypkin commented on x-pack/plugins/security/public/account_management/index.ts:
```diff
@@ -6,3 +6,6 @@
 
 export { accountManagementApp } from './account_management_app';
+
+export type { ChangePasswordProps } from './change_password/change_password';
+export type { PersonalInfoProps } from './personal_info/personal_info';
```
__nit:__ please export these types from `./change_password | personal_info/index.ts`. We use `index.ts` as a grouping mechanism for all exports where feasible.

```suggestion
export type { ChangePasswordProps } from './change_password';
export type { PersonalInfoProps } from './personal_info';
```

yakhinvadim replied:
Got it, thanks! Updated here: https://github.com/elastic/kibana/pull/99791/commits/9b932dcf9ba0c4a2a65fe5ecd792f3db93cfebab

azasypkin commented on x-pack/plugins/security/public/ui_api/index.ts:
```diff
@@ -0,0 +1,34 @@
+export interface UiApi {
+  components: {
+    getPersonalInfo: LazyComponentFn<PersonalInfoProps>;
+    getChangePassword: LazyComponentFn<Pick<ChangePasswordProps, 'user'>>;
```
__nit:__ (once you split this one in two interfaces)
```suggestion
    getChangePassword: LazyComponentFn<ChangePasswordProps>;
```

yakhinvadim replied:
Updated here: https://github.com/elastic/kibana/pull/99791/commits/24258211021c0b995ff548521a172acaac11c636

azasypkin commented on x-pack/plugins/security/public/ui_api/components.tsx:
```diff
@@ -0,0 +1,35 @@
+
+import type { CoreStart } from 'src/core/public';
+
+import { getChangePasswordComponent } from '../account_management/change_password/change_password_async';
+import { getPersonalInfoComponent } from '../account_management/personal_info/personal_info_async';
```
__nit:__ (once you propagate all required exports through relevant `index.ts` files)
```suggestion
import { getChangePasswordComponent,  getPersonalInfoComponent } from '../account_management';
```

yakhinvadim replied:
I tried to did it before, but it resulted in an increased bundle size.

I took the following info from `security/target/public/metrics.json` after running the first command from [CI Metrics doc](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html#ci-metric-validating-limits).

Before:
```
{
  "group": "page load bundle size",
  "id": "security",
  "value": 87166,
  "limit": 95864,
  "limitConfigPath": "packages/kbn-optimizer/limits.yml"
}
```
  
After:
```  
{
  "group": "page load bundle size",
  "id": "security",
  "value": 101392,
  "limit": 95864,
  "limitConfigPath": "packages/kbn-optimizer/limits.yml"
}
```

I'm guessing that once async components are exported through the index file and we use that index file somewhere that also brings all the sync components' dependencies and increases bundle size.

I kept the imports as is, but let me know if you have more thoughts.

yakhinvadim replied:
Good idea, absolutely.

yakhinvadim replied:
Added a comment here: https://github.com/elastic/kibana/pull/99791/commits/1880755587bfed1c337da8cfc626999e52569041

azasypkin commented on x-pack/plugins/enterprise_search/public/applications/workplace_search/views/account_settings/account_settings.tsx:
```diff
@@ -0,0 +1,39 @@
+  const [currentUser, setCurrentUser] = useState<AuthenticatedUser | null>(null);
+
+  useEffect(() => {
+    security!.authc!.getCurrentUser().then(setCurrentUser);
```
__question:__ won't we have an unhandled exception here in case security plugin is disabled?

yakhinvadim replied:
Thanks for checking our plugin code as well! Good question.

No, if this component loads, we are guaranteed to have the security plugin enabled. 
One of the requirements for being able to run Workplace Search is to have security enabled (see Step 2 [in our installation docs](https://www.elastic.co/guide/en/workplace-search/current/workplace-search-install.html#installation-self-managed)). If security is disabled, Workplace Search won't run and this page could not be opened.

azasypkin replied:
Got it, thanks for clarifying! Was just confused by these `!` and was wondering if `security` and its properties can really be `undefined` or not (by the way it seems `!` isn't strictly needed for `security` itself and is only needed to access its properties).

yakhinvadim replied:
You're right. I guess I overdid it here. 😄 
I plan to do a follow-up PR with some cleanup in the enterprise_search plugin only.
I'll update this line there if you don't mind.

azasypkin replied:
Sounds good to me :+1:

azasypkin reviewed (approved):
LGTM, just few more nits, thanks :+1:

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/130509/)
* Commit: 1880755587bfed1c337da8cfc626999e52569041
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-99791/1880755587bfed1c337da8cfc626999e52569041)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/49ea272240cefe6689d675ebac5ca303a4650d67) | [after](https://github.com/elastic/kibana/commit/1880755587bfed1c337da8cfc626999e52569041) | diff |
| --- | --- | --- | --- |
| `enterpriseSearch` | 1418 | 1420 | +2 |
| `security` | 475 | 482 | +7 |
| total |  |  | +9 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/49ea272240cefe6689d675ebac5ca303a4650d67) | [after](https://github.com/elastic/kibana/commit/1880755587bfed1c337da8cfc626999e52569041) | diff |
| --- | --- | --- | --- |
| `enterpriseSearch` | 2.1MB | 2.1MB | +1.7KB |
| `security` | 766.7KB | 782.3KB | +15.6KB |
| total |  |  | +17.3KB |

#### Public APIs missing exports
> Total count of every type that is part of your API that should be exported but is not. This will cause broken links in the API documentation system. Target amount is 0. `Run node scripts/build_api_docs --plugin [yourplugin] --stats exports` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/49ea272240cefe6689d675ebac5ca303a4650d67) | [after](https://github.com/elastic/kibana/commit/1880755587bfed1c337da8cfc626999e52569041) | diff |
| --- | --- | --- | --- |
| `security` | 5 | 6 | +1 |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/49ea272240cefe6689d675ebac5ca303a4650d67) | [after](https://github.com/elastic/kibana/commit/1880755587bfed1c337da8cfc626999e52569041) | diff |
| --- | --- | --- | --- |
| `security` | 82.5KB | 85.3KB | +2.8KB |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/49ea272240cefe6689d675ebac5ca303a4650d67) | [after](https://github.com/elastic/kibana/commit/1880755587bfed1c337da8cfc626999e52569041) | diff |
| --- | --- | --- | --- |
| `security` | 109 | 110 | +1 |

#### async chunk count
| id | [before](https://github.com/elastic/kibana/commit/49ea272240cefe6689d675ebac5ca303a4650d67) | [after](https://github.com/elastic/kibana/commit/1880755587bfed1c337da8cfc626999e52569041) | diff |
| --- | --- | --- | --- |
| `security` | 21 | 23 | +2 |


</details>

### History
* :green_heart: [Build #129846](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/129846/) succeeded 140cd91fc7ea4717be957322a215833d0f9ba368
* :green_heart: [Build #128501](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128501/) succeeded efa7d683110ce4bd9b6533e1eddf7bf17f9a59ed
* :broken_heart: [Build #125561](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125561/) failed 2c350dc99b4f5d8f01718141cd882d57fa1ee96d
* :broken_heart: [Build #125333](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125333/) failed 48d7bfa568a395f9d405d347d40c1fffdeaf623d
* :broken_heart: [Build #125314](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125314/) failed 3c87b4b4a44a14c2478323c4e52aa00776e46b26

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125314/","number":"125314","commit":"3c87b4b4a44a14c2478323c4e52aa00776e46b26"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125333/","number":"125333","commit":"48d7bfa568a395f9d405d347d40c1fffdeaf623d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/125561/","number":"125561","commit":"2c350dc99b4f5d8f01718141cd882d57fa1ee96d"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/128501/","number":"128501","commit":"efa7d683110ce4bd9b6533e1eddf7bf17f9a59ed"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/129846/","number":"129846","commit":"140cd91fc7ea4717be957322a215833d0f9ba368"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/130509/","number":"130509","commit":"1880755587bfed1c337da8cfc626999e52569041"}]}
PIPELINE-->

scottybollinger reviewed (approved):
This is great work Vadim! Thanks for knocking it out 👍

yakhinvadim commented:
Thanks to everyone and especially @azasypkin for helping with this! 🙌

kibanamachine commented:
## 💚 Backport successful

| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [7.x](https://github.com/elastic/kibana/pull/101930)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/101930">](https://github.com/elastic/kibana/pull/101930) |

This backport PR will be merged automatically after passing CI.
